### ISSUE TITLE
Redis OCF is not working (for slaves)

### ISSUE THREAD

**Issue by lejeczek** [ClusterLabs/resource-agents#1646]
Hi.

With:
pacemaker-2.0.5-9.el8.x86_64
redis-6.0.9-2.module_el8.4.0+580+92c8284e.x86_64
redis-doc-6.0.9-2.module_el8.4.0+580+92c8284e.noarch
pcs-0.10.8-1.el8.x86_64

Redis cluster works as expected outside of pacemaker/pcs.
I try to follow OCF_HEARTBEAT_REDIS man page which suggest setting Redis will be simple.
I do:

`-> $ pcs resource create redis ocf:heartbeat:redis bin=/usr/bin/redis-server client_bin=/usr/bin/redis-cli config=/etc/redis.conf rundir=/run/redis user=redis op monitor interval="1s" role="Master" timeout="5s" on-fail="restart" op monitor interval="2s" role="Slave" timeout="5s" on-fail="restart" promotable`

Master starts on the node cluster chooses:
...
`1281452:M 08 May 2021 09:09:46.423 * MASTER MODE enabled (user request from 'id=15 addr=/run/redis/redis.sock:0 fd=9 name= age=0 idle=0 flags=U db=0 sub=0 psub=0 multi=-1 qbuf=34 qbuf-free=32734 argv-mem=12 obl=0 oll=0 omem=0 tot-mem=61476 events=r cmd=slaveof user=default'`

but slaves, though cluster reports:
...
`* Clone Set: redis-clone [redis] (promotable):`
`* redis (ocf::heartbeat:redis): Slave 10.1.1.224`
`* redis (ocf::heartbeat:redis): Master 10.1.1.223`
`* redis (ocf::heartbeat:redis): Slave 10.1.1.225`

fail with this in the log:
...
`# Unable to connect to MASTER: Resource temporarily unavailable`

What is more interesting and puzzling - on such PCS/RA managed cluster for redis slave do:

`-> $ pcs resource ban redis-clone 10.1.1.224; pcs resource clear redis-clone 10.1.1.224`

and that "failed" node connects to the master in that instant.
Also, moving master clone of such "resurrected" cluster to another node also fails.

many thanks, L.

**Comment by dciabrin** [comment]
Looking into it.
FTR, we're using it in our OpenStack deployments, so I'll check what we may do differently from what's reported here.

Also, to clear out any ambiguities, the Master/Slave tag as reported in pacemaker represents the status of the pacemaker resource. M/S resources are two-state resources, so when they are started, they are either slave or master. In the case of this resource Master should match the redis state, and so is slave.

**Comment by lejeczek** [comment]
Yes, that is what is see pacemaker is doing, what it deems as master-clone appears to be Redis master and respectively for the slaves.
I mistakenly reported it under 'PCS' and @tomjelinek said:
"
This is just a guess:
It seems that when a slave instance is restarted (by ban and clear), it starts working just fine. If I'm not mistaken, pacemaker starts resources in slave mode and then chooses one instance and promotes it to master mode. Maybe the slaves are connecting to the master when there is no master yet?
"

**Comment by dciabrin** [comment]
Ok I think I reproduced it with a simple redis setup...
pcs resource create redis-clone  ocf:heartbeat:redis promotable meta master=1

I can indeed see that the slave are not connected to the master after the bootstrap and the monitor function is not enough to force the slave to reconnect. A resource restart will, but that's still not what we want.

I'll have a look at what's going on shortly, thanks for reporting it.

**Comment by dciabrin** [comment]
Heads up, this was a false alarm.

The way the redis resource agent work is by reacting to master promotion to connect to it. So in fact, the redis resource must be created with notification enabled, for the right code path to be triggered:

pcs resource create redis ocf:heartbeat:redis promotable notify=true ordered=true

We also use ordered=true in our Openstack deployments, so the start of the redis cluster os orchestrated rather than fully parallel, which makes no improvement on startup time.

When created like than, you should see right notification in the journal telling you the slave is connecting to the new master:

[At start, when no master have be created just yet]
May 19 10:21:23 rhel3 redis(redis)[1690808]: INFO: monitor: Slave mode link has not yet been established (link=down)
May 19 10:21:23 rhel3 redis(redis)[1690855]: INFO: demote: Setting master to 'no-such-master'

[Once a new master is promoted, e.g. on rhel2 in this case] 
May 19 10:21:27 rhel3 redis(redis)[1691293]: INFO: monitor: Slave mode link has not yet been established (link=down)
May 19 10:21:27 rhel3 redis(redis)[1691340]: INFO: demote: Setting master to 'rhel2'

**Comment by lejeczek** [comment]
I would urge your if you own/maintain, or urge you to urge whoever is responsible for, man pages, to include and explain those **critical** bits in man for OCF_HEARTBEAT_REDIS
Should not have to say how many of us regular folks turned away too quickly after stumbling upon this "issue".
many thanks, L.
